### PR TITLE
[SPARK-26580][SQL] remove Scala 2.11 hack for Scala UDF

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In https://github.com/apache/spark/pull/22732 , we tried our best to keep the behavior of Scala UDF unchanged in Spark 2.4.

However, since Spark 3.0, Scala 2.12 is the default. The trick that was used to keep the behavior unchanged doesn't work with Scala 2.12.

This PR proposes to remove the Scala 2.11 hack, as it's not useful.

## How was this patch tested?

existing tests.

### CODE REVIEW

File: docs/sql-migration-guide-upgrade.md [hunk 1]
```diff
@@ -43,6 +43,8 @@ displayTitle: Spark SQL Upgrading Guide
 
   - Since Spark 3.0, JSON datasource and JSON function `schema_of_json` infer TimestampType from string values if they match to the pattern defined by the JSON option `timestampFormat`. Set JSON option `inferTimestamp` to `false` to disable such type inferring.
 
+  - In Spark version 2.4 and earlier, if `org.apache.spark.sql.functions.udf(Any, DataType)` gets a Scala closure with primitive-type argument, the returned UDF will return null if the input values is null. Since Spark 3.0, the UDF will return the default value of the Java type if the input value is null. For example, `val f = udf((x: Int) => x, IntegerType)`, `f($"x")` will return null in Spark 2.4 and earlier if column `x` is null, and return 0 in Spark 3.0.
```
Comment (cloud-fan) [inline docs/sql-migration-guide-upgrade.md]: this migration guide should have been added when we switch to Scala 2.12.

**Comment by cloud-fan** [comment]
cc @maryannxue @srowen @gatorsmile @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Yea, I agree with this change.

**Review by srowen — APPROVED** [review]

File: docs/sql-migration-guide-upgrade.md (hunk 1 above)
Comment (felixcheung) [reply docs/sql-migration-guide-upgrade.md]: should this say this is because of Scala 2.12?

**Comment by cloud-fan** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**